dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordVectorPropertyModel.cs

```diff
@@ -54,6 +60,147 @@ public int Dimensions
+    /// and <see cref="DataContent"/> as input types for <see cref="EmbeddingGenerator"/>.
+    /// </summary>
+    public virtual bool TrySetupEmbeddingGeneration<TEmbedding, TUnwrappedEmbedding>(IEmbeddingGenerator embeddingGenerator, Type? embeddingType)
+        where TEmbedding : Embedding
+    {
+        // On the TInput side, this out-of-the-box/simple implementation supports string and DataContent only
+        // (users who want arbitrary TInput types need to use the generic subclass of this type).
+        // The TEmbedding side is provided by the connector via the generic type parameter to this method, as the connector controls/knows which embedding types are supported.
+        // Note that if the user has manually specified an embedding type (e.g. to choose Embedding<Half> rather than the default Embedding<float>), that's provided via the embeddingType argument;
+        // we use that as a filter below.
+        switch (embeddingGenerator)
+        {
+            case IEmbeddingGenerator<string, TEmbedding> when this.Type == typeof(string) && (embeddingType is null || embeddingType == typeof(TUnwrappedEmbedding)):
+            case IEmbeddingGenerator<DataContent, TEmbedding> when this.Type == typeof(DataContent) && (embeddingType is null || embeddingType == typeof(TUnwrappedEmbedding)):
+                this.EmbeddingGenerator = embeddingGenerator;
+                this.EmbeddingType = embeddingType ?? typeof(TUnwrappedEmbedding);
+
+                return true;
+
+            case null:
+                throw new UnreachableException("This method should only be called when an embedding generator is configured.");
+            default:
+                return false;
+        }
+    }
+
+    /// <summary>
+    /// Attempts to generate an embedding of type <typeparamref name="TEmbedding"/> from the vector property represented by this instance on the given <paramref name="record"/>, using
+    /// the configured <see cref="EmbeddingGenerator"/>.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// If <see cref="EmbeddingGenerator"/> supports the given <typeparamref name="TEmbedding"/>, returns <see langword="true"/> and sets <paramref name="task"/> to a <see cref="Task"/>
+    /// representing the embedding generation operation. If <see cref="EmbeddingGenerator"/> does not support the given <typeparamref name="TEmbedding"/>, returns <see langword="false"/>.
+    /// </para>
+    /// <para>
+    /// The implementation on this non-generic <see cref="VectorStoreRecordVectorPropertyModel"/> checks for <see cref="string"/>
+    /// and <see cref="DataContent"/> as input types for <see cref="EmbeddingGenerator"/>.
+    /// </para>
+    /// </remarks>
+    public virtual bool TryGenerateEmbedding<TRecord, TEmbedding, TUnwrappedEmbedding>(TRecord record, CancellationToken cancellationToken, [NotNullWhen(true)] out Task<TEmbedding>? task)
```

Comment:
As I wrote in the other comment, having a public method that returns a `Task` via `out` argument is very odd to me.

The alternatives I can see right now (please bear in mind I am before my first coffee):
- return a `ValueTask<TEmbedding?>` and have sth like `var enmbedding = await TryGenerateEmbedding<float>() ?? await TryGenerateEmbedding<double>() ?? throw InvalidOperationException`
- have  a public helper method that takes care of all the properties in given record (the code that iterates over `_model.VectorProperties` and calls this method) and simply don't expose `TryGenerateEmbedding` as public

Reply:
Yeah, I definitely want to revisit all of this with a fresh mind after the release pressure is off.